Step3

ivvve opened the pull request:
안녕하세요. 리뷰이인 손영철입니다!
좋은 주말 보내셨나요??

이번주에 회사일도 많았지만 심적으로 좀 요동쳐서 리뷰요청이 좀 늦었습니다...

그리고 Step3를 진행하면서 몇가지 질문할 것이 있는데요.

첫째로, 
RestSecurityControllerAdvice에서 원래 ErrorMessage를 리턴하게 되어있는데 문제가 나서 현재는 아래와 같이 주석으로 return을 막아놓았는데요.

```java
@ExceptionHandler(UnAuthenticationException.class)
@ResponseStatus(value = HttpStatus.UNAUTHORIZED)
public void unAuthentication(UnAuthenticationException e) {
    log.debug("JSON API UnAuthenticationException is happened!");
//        return new ErrorMessage(e.getMessage());
}
```

기존 코드에서 처럼 return을 하면 API Acceptance 테스트 시 login이 필요한 경우에 대해 테스트를 할 때 관련되서 Exception이 발생해서 테스트가 깨집니다.

```
(Exception 내용은 길어서 중간은 생략하였습니다.)
12:59:43.755 [DEBUG] [http-nio-auto-1-exec-3] [n.security.BasicAuthInterceptor] - Authorization : null
12:59:43.756 [DEBUG] [http-nio-auto-1-exec-3] [n.s.RestSecurityControllerAdvice] - JSON API UnAuthenticationException is happened!
12:59:43.820 [WARN ] [http-nio-auto-1-exec-3] [o.s.w.s.m.m.a.ExceptionHandlerExceptionResolver] - Failure in @ExceptionHandler public support.domain.ErrorMessage nextstep.security.RestSecurityControllerAdvice.unAuthentication(nextstep.UnAuthenticationException)
org.springframework.web.HttpMediaTypeNotAcceptableException: Could not find acceptable representation
...
12:59:43.822 [ERROR] [http-nio-auto-1-exec-3] [o.a.c.c.C.[.[.[.[dispatcherServlet]] - Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is nextstep.UnAuthenticationException: You're required Login!] with root cause
nextstep.UnAuthenticationException: You're required Login!
	at nextstep.security.LoginUserHandlerMethodArgumentResolver.resolveArgument(LoginUserHandlerMethodArgumentResolver.java:27)
```

하지만 application을 실행시키고 문제가 되는 부분을 실행하였을 때는 이상이 없었서 혹시 해당 이슈에 대해 묻고싶습니다.

---

둘째로,
개별적으로 테스트를 실행하면 문제가 없는데 테스트 코드를 같이 돌리면 깨지는 경우가 생깁니다.
ApiQuestionAcceptanceTest에서 147번째 라인을 보면 삭제한 Question을 다시 원복시키는 메서드를 실행하게 해놓았는데

```java
    @Test
    public void question_delete_login() {
        // given
        User loginUser = defaultUser();
        Question question = defaultQuestion();

        // when
        ResponseEntity<String> response = deleteQuestionResource(loginUser, question);

        // then
        Question dbQuestion = getQuestionResourceWithoutLogin(question.generateRestUrl()).getBody();

        softly.assertThat(response.getStatusCode()).isEqualByComparingTo(HttpStatus.NO_CONTENT);
        softly.assertThat(dbQuestion.isDeleted()).isTrue();

        log.debug("response body : {}", response.getBody());

        // tearDown
        restoreDeletedQuestion(dbQuestion);
    }
```

이유가 해당 테스트가 다른 테스트에 영향을 주어서 다른 테스트가 깨져서입니다.
테스트 메서드 마다 다른 Question을 가져와서 테스트하는 식으로 테스트를 짜야할까요?
해당 문제에 대한 의견을 묻고 싶습니다.
(@After를 사용하지 않은 이유는 특정 테스트에서만 해당 이슈가 발생하기 해서 전체 테스트에 적용할 필요가 없을 것 같다고 판단되서입니다.)

---

개발하고 다시 하기를 여러번해서 몇일동안 진행하다보니 로직이 이상하거나 주석처리했는데 제가 발견하지 못한 부분이 있을 것 같습니다. ㅠㅠ
문제 있는 부분 발견하시면 말씀해주시면 바로 반영하도록 하겠습니다!
그러면 리뷰요청드리겠습니다.

감사합니다!

ivvve committed:
질문 수정, 삭제 시 자신이 쓴 글인지 확인하는 책임을 Question 객체에 위임

ivvve committed:
중복 코드 메서드화

ivvve committed:
공통 메서드 변경

ivvve committed:
AcceptanceTest 공통 메서드 로직 변경

ivvve committed:
AcceptanceTest 공통 메서드 추가 : delete

ivvve committed:
라인 정리

ivvve committed:
client에서 RestAPI로 잘못된 parameter를 보냈을 때 발생하는 IllegalArgument를 처리하기 위한 ControllerAdvice

ivvve committed:
ErrorMessage를 return하면 테스트 시
org.springframework.web.HttpMediaTypeNotAcceptableException가
발생해서 void로 처리해놓음

ivvve committed:
1. delete, update 메서드 추가
2. api url 변경

ivvve committed:
update 후 해당 객체 return 하도록 변경

ivvve committed:
Answer CRUD 추가

ivvve committed:
Answer REST API : CRD 구현

ivvve committed:
Answer REST API : update 구현

ivvve committed:
Question 유닛테스트 생성

ivvve committed:
테스트 로직 변경

ivvve committed:
1. restUrl 생성 추가
2. 생성자 추가

ivvve committed:
add Answer unit test

ivvve committed:
Question API 구현

ivvve committed:
update, delete시 삭제된 질문에 대한 처리 로직 추가

ivvve committed:
Answer API 구현

ivvve committed:
answer id와 question id로 Answer를 가져오는 메서드 추가

ivvve committed:
generateUrl 로직 변경

ivvve committed:
테스트명 변경

kang-hyungu reviewed (changes requested):
안녕하세요 영철님
회사일로 많이 바쁘셨나보네요 ㅠㅠ
4기 과정 끝나고 pr 주셔도 괜찮습니다.
천천히 리뷰 요청 주셔도 되니 정리 되시면 차근차근 수정해보아요~!

1. 깨지는 테스트 확인해주세요!
ApiAnswerAcceptanceTest.answer_update_no_login()
QuestionAcceptanceTest.question_update_login_작성자()

2. RestSecurityControllerAdvice에서 수정하신 부분 원복하고 테스트 돌렸을 때 ApiAnswerAcceptanceTest.answer_create_no_login 테스트에서 말씀하셨던 에러가 발생하네요.
UnAuthenticationException: You're required Login! 라는 에러 메시지를 보니 로그인 안한 사용자가 답변을 추가할 때 발생하는 메시지네요. 로그인해야 답변 가능하므로 정상 메시지이로 보이고 exception 테스트를 추가하시면 될 것 같네요.
이외에 다른 부분에서도 위와 같은 에러 발생한다면 어떤 테스트인지 코멘트 남겨주세요.

3. 다른 테스트에 영향을 받지 않도록 테스트 코드를 작성하시는게 좋습니다.
@before와 @after를 활용하셔도 좋고 테스트 수행 전에 질문 생성 api를 호출하고 api 응답값에 질문 번호로 삭제하면 다른 테스트에 영향을 주지 않을거에요.
인수 테스트는 사용자 입장에서 제품이 정상 동작하는지 확인하는 테스트인만큼 repository에서 직접 조회하기보다 완성된 api로 정상 동작하는지 실제 사용자가 사용하는 것처럼 테스트 코드 작성하는게 좋습니다.
인수 테스트에서 repository를 제거해보는건 어떨까요

세 가지 피드백 드린 부분 수정해보시기 바랍니다!

ivvve committed:
createAnswer : form-urlencoded에서 application/json으로 받아오는 것으로 변경

ivvve committed:
Queston의 cascade로 인한 중복 로직 제거

ivvve committed:
1. 테스트 시 새로 Answer를 생성하여 테스트끼리 서로 영향을 주지 않도록 변경
2. AnswerRepository를 통해서 변경된 Answer를 가져오지 않고 API를 호출하여 해당 Answer를 가져오도록 변경

ivvve committed:
defaultAnswer 삭제

ivvve committed:
1. 테스트 시 새로 Question을 생성하여 테스트끼리 서로 영향을 주지 않도록 변경
2. QuestionRepository를 통해서 변경된 Question을 가져오지 않고 API를 호출하여 해당 Question을 가져오도록 변경

ivvve committed:
기존 defaultUser외에 'sanjigi' user를 가져오는 메서드 추가

ivvve committed:
1. 테스트 시 새로 Question을 생성하여 테스트끼리 서로 영향을 주지 않도록 변경

ivvve committed:
resource가 존재하지 않을 때 IllegalArgumentException -> NoSuchElementException을 throw 하도록 변경

ivvve committed:
기존 코드로 원복

ivvve commented:
안녕하세요 현구님!
오늘 하루 잘 보내셨나요? :)

이번주는 업무양도 그렇고 멘탈도 저번주보다 좋아져서 다시 즐겁게 과제를 했습니다.

먼저 저번에 주신 피드백 감사드립니다!
저번에 주신 피드백 반영하여 코드 수정하여 다시 리뷰요청드립니다.

---

1. 피드백을 참고해서 수정한 REST API AcceptanceTest가 기존보다 좀 더 간결하고 깔끔해진 것 같습니다. 👍 

2. 기존에 RestSecurityControllerAdvice 변경했던 코드를 다시 원복시켰는데 이전의 exception이 일어나지 않네요;;
정상적인 결과 같은데 이전에 일어나던 오류가 다시 나타나지 않아 당황스럽네요;;

3. 그리고 하나 물어볼 것이 있습니다!
Acceptance test에서 repository를 빼보라고 하셨는데
REST API의 경우 API 호출을 통해 resource를 가져올 수 있는데
QuestionAcceptanceTest의 경우 해당 컨트롤러를 통해 resource를 가져올 수 없어 repository를 사용했습니다.
이렇게 REST API가 아니고 view를 이용한 controller의 경우는 어떤식으로 테스트 하는 쪽이 좋을까에 대한 의견을 듣고싶습니다!

---

그럼 리뷰 요청 드리겠습니다.
감사합니다!

kang-hyungu reviewed (approved):
피드백 반영 확인했습니다!
질문 주신 QuestionAcceptanceTest 같이 view를 사용하는 경우는 HtmlUnit이나 selenium을 사용하는 방법도 있습니다. 하지만 테스트 작성이 어렵고 화면단은 변경이 무척 자주 발생하기 때문에 테스트 코드 유지가 어려워 잘 보통 테스트 작성을 잘 안하게 됩니다; 요즘엔 SPA 방식으로 개발하다보니 프론트와 백단 코드가 분리되서 api 테스트에 집중할 수 있더라구요.
QuestionAcceptanceTest 같은 경우는 repository를 사용하는게 코드도 깔끔하고 유지보수에 더 유리하다고 보여서 굳이 제거할 필요는 없어보여요. 이번 과제는 api 개발 및 테스트 추가라서 관련 테스트들만 말씀드린거였어요.

구현하신 내용은 머지할게요. 마지막 단계 진행하시기 바랍니다
수고하셨습니다!
